state-chain/pallets/cf-swapping/src/lib.rs

```diff
@@ -834,64 +882,83 @@ pub mod pallet {
 
 				*stable_amount = remaining_amount;
 
+				// Copy so we don't hold a mutable reference:
+				let stable_amount = *stable_amount;
+
 				Self::deposit_event(Event::<T>::NetworkFeeTaken {
 					fee_amount: network_fee,
-					swap_id: swap.swap_id,
+					swap_id: swap.swap_id(),
 				});
 
-				if swap.to == STABLE_ASSET {
-					swap.final_output = Some(*stable_amount);
+				if swap.output_asset() == STABLE_ASSET {
+					swap.final_output = Some(stable_amount);
 				}
 			}
 
 			Ok(())
 		}
 
-		// Transactional ensures that any failed swap will rollback all storage changes.
 		#[transactional]
-		fn process_swaps_for_block(block: BlockNumberFor<T>) -> Result<(), BatchExecutionError> {
-			let mut swaps = SwapQueue::<T>::take(block);
-
-			if swaps.is_empty() {
-				return Ok(())
-			}
+		fn execute_batch(swaps: Vec<Swap>) -> Result<Vec<SwapCtx>, BatchExecutionError> {
+			let mut swaps: Vec<_> = swaps.into_iter().map(SwapCtx::new).collect();
 
 			// Swap into Stable asset first, then take network fees:
 			Self::swap_into_stable_taking_network_fee(&mut swaps)?;
 
 			// Swap from Stable asset, and complete the swap logic.
 			Self::do_group_and_swap(&mut swaps, SwapLeg::FromStable)?;
 
+			// Swaps executed without triggering price impact protection, but we still need to
+			// check that none of the swaps violated their minimum output requirements:
+			let (non_violating, violating): (Vec<_>, Vec<_>) =
+				swaps.into_iter().partition(|swap| {
+					let final_output = swap.final_output.unwrap();
+					swap.refund_params()
+						.as_ref()
+						.map_or(true, |params| final_output >= params.min_output)
+				});
+
+			if violating.is_empty() {
+				Ok(non_violating)
+			} else {
+				Err(BatchExecutionError::PriceLimitHit {
+					successful_swaps: non_violating.into_iter().map(|ctx| ctx.swap).collect(),
+					failed_swaps: violating.into_iter().map(|ctx| ctx.swap).collect(),
+				})
+			}
+		}
+
+		fn process_swap_outcomes(swaps: &[SwapCtx]) {
 			for swap in swaps {
 				if let Some(swap_output) = swap.final_output {
 					// To be consistent with `swap_output` and `intermediate_amount` (which do
 					// not include the network fee), we report input amount without the network fee
 					// for swaps from STABLE_ASSET:
-					let swap_input = if swap.from == STABLE_ASSET {
+					let swap_input = if swap.input_asset() == STABLE_ASSET {
```

Comment:
ik this was the case before, but perhaps something to consider, it does feel strange that we have this swap_input() calculation done here and not inside the SwapCtx, which I would expect to handle this case. We have `swap_amount` - which is used twice and does an `unwrap_or_default()` to remove the option in each case. Is just indicative to me that something can be improved here (nothing you necessarily need to address here, but food for thought)